packages/tiny-router/src/actions.ts

```diff
@@ -83,6 +90,20 @@ export const init: TinyRouter["actions"]["router"]["init"] = ({
         ? libraries.source.normalize(state.frontity.initialLink)
         : state.frontity.initialLink;
   } else {
+    // Observe the current data object. If it is ever a redirection, replace the
+    // current link with the new one.
+    observe(() => {
+      const data = state.source?.get(state.router.link);
+      if (data && isRedirection(data)) {
+        actions.router.set(data.location, {
+          // Use "replace" to keep browser history consistent.
+          method: "replace",
+          // Keep the same history.state that the old link had.
+          state: state.router.state,
```

Comment:
@luisherranz I've discovered that passing the state back to the router will fail if the state is already proxified.

It seems that you cannot pass a `Proxy` object as the `state` parameter, you can try it yourself:

```
> const proxy = new Proxy({}, {})
undefined
> window.history.replaceState(p, "", window.location.href);
VM178:1 Uncaught DOMException: Failed to execute 'replaceState' on 'History': [object Object] could not be cloned.
    at <anonymous>:1:16
```

I think we have two options. We could either "clone" the state object as suggested in https://stackoverflow.com/questions/53102700/how-do-i-turn-an-es6-proxy-back-into-a-plain-object-pojo but I think that a better solution would be to add an "escape" hatch like discussed in https://stackoverflow.com/a/60668790/2638310. We add a special symbol to the state which returns the "original", unproxied object. Like:

```
const original = {foo: 'bar'};

const handler = {
  get(target, prop) {
    if (prop === Symbol.for('ORIGINAL')) return target;
    return Reflect.get(target, prop);
  }
};

const proxy = new Proxy(original, handler);
const unproxied = proxy[Symbol.for('ORIGINAL')];

console.assert(original === unproxied); // true
```

I think that this should be an implementation detail for now, but maybe in the future we could expose a more ergonomic API for this in `@frontity/connect`

What do you think?

Reply:
You are absolutely right.

@orballo already stumbled upon this problem in the `infinite-scroll-beta` branchan and we decided to go with the `clone` approach because it is safer.

`react-easy-state`/`@frontity/connect` has a [`raw()` utility to expose the original object](https://github.com/frontity/frontity/blob/dev/packages/connect/src/index.js#L2), but it is not deep, so would have to create a `deepRaw()` that iterates over the object extracting all the originals, in case there are proxies inside proxies.

So we decided to go with the `clone` approach for now, as you can see in the `actions.ts` file of `tiny-router` of the `infinite-scroll-beta` branch: https://github.com/frontity/frontity/blob/infinite-scroll-beta/packages/tiny-router/src/actions.ts.